src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
-                        @info "Accept ($i):  $conn"
-                        connectioncounter[] += 1
-                        handle_connection(f, conn; kw...)
+"main server loop that accepts new tcp connections and spawns async threads to handle them"
+function listenloop(h, server,
+    tcpisvalid=x->true, connectioncounter=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
+    count = 1
+    while isopen(server)
+        try
+            io = accept(server)
+            io === nothing && continue
+            if !tcpisvalid(io)
+                verbose && @info "Accept-Reject:  $io"
+                close(io)
+                continue
+            end
+            connectioncounter[] += 1
+            conn = Connection(server.hostname, server.hostport, 0, 0, true, io)
+            let io=io, count=count
+                @async begin
+                    try
+                        verbose && @info "Accept ($count):  $conn"
+                        handle(h, conn, reuse_limit, readtimeout)
+                        verbose && @info "Closed ($count):  $conn"
                     catch e
-                        @error "Error ($i):  $conn" exception=(e, stacktrace(catch_backtrace()))
+                        @error exception=(e, stacktrace(catch_backtrace()))
                     finally
                         connectioncounter[] -= 1
-                        close(conn)
-                        @info "Closed ($i):  $conn"
+                        close(io)
+                        verbose && @info "Closed ($count):  $conn"
                     end
                 end
-            catch e
-                if e isa Base.IOError
-                    @warn "$e"
-                    break
-                else
-                    rethrow(e)
-                end
             end
-            id += 1
-        end
-    catch e
-        if e isa InterruptException
-            @warn "Interrupted: listen($hostname)"
-        else
-            rethrow(e)
+        catch e
+            if e isa InterruptException
+                @warn "Interrupted: listen($server)"
+                close(server)
+                break
+            end
+            @error exception=(e, stacktrace(catch_backtrace()))
         end
-    finally
-        close(tcpserver)
+        count += 1
     end
-
     return
 end
 
-
 """
-Start a timeout monitor task to close the `Connection` if it is inactive.
-Create a `Transaction` object for each HTTP Request received.
+Connection handler: starts an async readtimeout thread if needed, then creates
+Transactions to be handled as long as the Connection stays open. Only reuse_limit + 1
+# of Transactions will be allowed during the lifetime of the Connection.
 """
-function handle_connection(f::Function, c::Connection;
-                           reuse_limit::Int=nolimit,
-                           readtimeout::Int=0, kw...)
+function handle(h, c::Connection,
+                reuse_limit::Int=10,
```

Comment:
As far as I can see this is only ever called from line 263. In which case the default values `10` and `0` are never used. Remove the defaults?

I would really rather not have this be a method of the global `handle` function. It seems like the intention of the `handle` function is for people to define custom methods and make generic calls.
I can't imagine any case where it would be a good idea to call this from somewhere else, or where a generic call to `handle` would end up passing a `HTTP.ConnectionPool.Connection`.

`Connection` is a private internal type. Even `Transaction` is not exposed directly (it is wrapped by `HTTP.Stream`). No one should be doing `http.stream.c` to get a `Connection` object and then calling `handle` on it. That would cause state to be messed up for other code that is not expecting it.

Can we keep the `handle_connection` name? It shouldn't make any user-visible difference right?